### PR TITLE
Remove MVP and add MVVM template code

### PR DESCRIPTION
**Pull request by mirland**

### CODE REVIEW

File: app/src/main/java/com/xmartlabs/template/App.kt [hunk 1]
```diff
@@ -82,17 +81,20 @@ open class App : Application(), HasActivityInjector {
...
-
   private fun initializeThreeTenABP() = AndroidThreeTen.init(this)
 
   private fun initializeLogging() {
     //TODO: Configure Fabric and add Fabric apiSecret and apiKey properties file in the root folder
     loggerTree.addLogger(CrashlyticsLogger().initialize(buildInfo, this))
     Timber.plant(loggerTree)
+
+    if (buildInfo.isDebug) {
+      Stetho.initializeWithDefaults(this);
```
Comment (bryant1410) [inline app/src/main/java/com/xmartlabs/template/App.kt]: Have you tried Steho? is it good?

File: app/src/main/java/com/xmartlabs/template/viewmodel/ViewModelFactory.kt [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+        val creator = creators[modelClass] ?: creators.entries.firstOrNull {
+            modelClass.isAssignableFrom(it.key)
+        }?.value ?: throw IllegalArgumentException("unknown model class $modelClass")
+        try {
+            @Suppress("UNCHECKED_CAST")
+            return creator.get() as T
+        } catch (e: Exception) {
+            throw RuntimeException(e)
+        }
+
```
Comment (f7deleon) [inline app/src/main/java/com/xmartlabs/template/viewmodel/ViewModelFactory.kt]: rm

**Review by f7deleon — APPROVED** [review]

File: app/src/main/java/com/xmartlabs/template/database/AppDataBase.kt [hunk 3]
```diff
@@ -1,10 +1,10 @@
...
-import com.raizlabs.android.dbflow.annotation.Database
+import android.arch.persistence.room.RoomDatabase
 
-@Database(name = AppDataBase.NAME, version = AppDataBase.VERSION)
-object AppDataBase {
-  //TODO: change db name
-  const val NAME = "template_database"
-  const val VERSION = 1
-}
+//@Database(
```
Comment (bryant1410) [inline app/src/main/java/com/xmartlabs/template/database/AppDataBase.kt]: Remove the commented code, or if it's gonna be used then add a TODO.

File: app/src/main/java/com/xmartlabs/template/di/ApplicationComponent.kt [hunk 4]
```diff
@@ -19,11 +19,13 @@ import javax.inject.Singleton
   AndroidModule::class,
   AppModule::class,
   ControllerModule::class,
+  DbModule::class,
   GsonModule::class,
   OkHttpModule::class,
   PicassoModule::class,
   RestServiceModule::class,
   RestServiceModuleApi::class,
+  ViewModelModule::class,
```
Comment (bryant1410) [inline app/src/main/java/com/xmartlabs/template/di/ApplicationComponent.kt]: alpha

File: app/src/main/java/com/xmartlabs/template/di/ViewModelKey.kt [hunk 5]
```diff
@@ -0,0 +1,15 @@
+package com.xmartlabs.template.di
+
+import android.arch.lifecycle.ViewModel
+import dagger.MapKey
+import kotlin.reflect.KClass
+
+@MustBeDocumented
```
Comment (bryant1410) [inline app/src/main/java/com/xmartlabs/template/di/ViewModelKey.kt]: why so? And shouldn't this class be public then?

File: app/src/main/java/com/xmartlabs/template/extensions/ui/ViewPagerExtensions.kt [hunk 6]
```diff
@@ -0,0 +1,20 @@
+@file:Suppress("NOTHING_TO_INLINE")
```
Comment (bryant1410) [inline app/src/main/java/com/xmartlabs/template/extensions/ui/ViewPagerExtensions.kt]: why not inlining?

File: app/src/main/java/com/xmartlabs/template/extensions/ui/ViewPagerExtensions.kt [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+
+inline fun ViewPager.nextPage() {
+  currentItem++
+}
+
+inline fun ViewPager.isLastPage(): Boolean {
+  return currentItem == (adapter?.let { adapter -> adapter.count - 1 })
+}
+
+inline fun ViewPager.onPageSelected(crossinline action: (position: Int) -> Unit) {
```
Comment (bryant1410) [inline app/src/main/java/com/xmartlabs/template/extensions/ui/ViewPagerExtensions.kt]: crossinline?

File: app/src/main/res/values/strings.xml [hunk 8]
```diff
@@ -1,7 +1,7 @@
 <resources>
     <string name="app_name" translatable="false">Template</string>
     <!--TODO: Replace base url with the correct url-->
-    <string name="base_url" translatable="false">https://xmartlabs.com/</string>
+    <string name="base_url" translatable="false">https://api.github.com/</string>
```
Comment (bryant1410) [inline app/src/main/res/values/strings.xml]: does this change matter?

File: app/src/sharedTest/java/com/xmartlabs/template/repository/MockAuthController.kt [hunk 9]
```diff
@@ -0,0 +1,10 @@
+package com.xmartlabs.template.repository
+
+import com.xmartlabs.template.service.AuthService
+import io.reactivex.Completable
+import javax.inject.Inject
+
+class MockAuthController @Inject constructor(authService: AuthService, sessionRepository: SessionRepository)
```
Comment (bryant1410) [inline app/src/sharedTest/java/com/xmartlabs/template/repository/MockAuthController.kt]: MockAuthRepository?

File: app/src/test/java/com/xmartlabs/template/common/TestComponent.kt [hunk 10]
```diff
@@ -33,6 +34,7 @@ import javax.inject.Singleton
   PicassoModule::class,
   RestServiceModule::class,
   RestServiceModuleApi::class,
+  ViewModelModule::class,
```
Comment (bryant1410) [inline app/src/test/java/com/xmartlabs/template/common/TestComponent.kt]: alpha

**Review by bryant1410 — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/xmartlabs/template/App.kt (hunk 1 above)
Comment (matir91) [reply app/src/main/java/com/xmartlabs/template/App.kt]: yep, it's good.

Comment (bryant1410) [reply app/src/main/java/com/xmartlabs/template/App.kt]: but what it's a good feature of it?

File: app/src/main/java/com/xmartlabs/template/database/AppDataBase.kt (hunk 3 above)
Comment (matir91) [reply app/src/main/java/com/xmartlabs/template/database/AppDataBase.kt]: I think todo is fine

File: app/src/main/java/com/xmartlabs/template/di/ViewModelKey.kt (hunk 5 above)
Comment (matir91) [reply app/src/main/java/com/xmartlabs/template/di/ViewModelKey.kt]: I think we can remove it, it's a copy paste from the google example

File: app/src/main/res/values/strings.xml (hunk 8 above)
Comment (matir91) [reply app/src/main/res/values/strings.xml]: no, I should remove this change

File: app/src/sharedTest/java/com/xmartlabs/template/repository/MockAuthController.kt (hunk 9 above)
Comment (matir91) [reply app/src/sharedTest/java/com/xmartlabs/template/repository/MockAuthController.kt]: yep

**Review by FerrariBruno — APPROVED** [review]
I'm approving for now, I'll review this later, when I have time.

File: app/src/main/java/com/xmartlabs/template/extensions/ui/ViewPagerExtensions.kt (hunk 6 above)
Comment (f7deleon) [reply app/src/main/java/com/xmartlabs/template/extensions/ui/ViewPagerExtensions.kt]: we use `inline fun` after that and the linter shows a warning with the methods if the suppression is not define

File: app/src/main/java/com/xmartlabs/template/App.kt (hunk 1 above)
Comment (matir91) [reply app/src/main/java/com/xmartlabs/template/App.kt]: It provides a simple way of inspect the network queries and the database queries

File: app/src/main/java/com/xmartlabs/template/extensions/ui/ViewPagerExtensions.kt (hunk 7 above)
Comment (matir91) [reply app/src/main/java/com/xmartlabs/template/extensions/ui/ViewPagerExtensions.kt]: https://android.jlelse.eu/inline-noinline-crossinline-what-do-they-mean-b13f48e113c2

File: app/src/main/java/com/xmartlabs/template/repository/common/NetworkState.kt [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+enum class Status {
+  RUNNING,
+  SUCCESS,
+  FAILED
+}
+
+@Suppress("DataClassPrivateConstructor")
+data class NetworkState private constructor(
+    val status: Status,
+    val throwable: Throwable? = null) {
```
Comment (f7deleon) [inline app/src/main/java/com/xmartlabs/template/repository/common/NetworkState.kt]: `)  {` goes in a new line

File: app/src/main/java/com/xmartlabs/template/repository/common/NetworkState.kt [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+  val msg: String?
+    get() = throwable?.message
+
+  companion object {
```
Comment (f7deleon) [inline app/src/main/java/com/xmartlabs/template/repository/common/NetworkState.kt]: shouldn't go before the `msg` val?

**Review by f7deleon — APPROVED** [review]

**Pull request merged**